### PR TITLE
Implements 1.8 codecs (0x41 up to 0x49) Partially completes #131

### PR DESCRIPTION
**Pull request by greatman**
Implement ServerDifficulty codec.
Implement CombatEvent codec.
Add 0x43 (Camera).
Packet 0x48 codec added.
Packet 0x44 added (WorldBorder)
Packet 0x45 added. (Title). Some formatting.
Add 0x49 (Entity update NBT). Using varLong now in WorldBorder.

Signed-off-by: William Lebel lebelwilliam@gmail.com

### CODE REVIEW

**Comment by greatman** [comment]
There should be good I think.

**Comment by turt2live** [comment]
Not sure what happened, but this PR causes problems with (at least) logging into the server. Haven't tested much more than that.

File: src/main/java/net/glowstone/net/codec/play/entity/UpdateEntityNBTCodec.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import com.flowpowered.networking.util.ByteBufUtils;
+import io.netty.buffer.ByteBuf;
+import net.glowstone.net.GlowBufUtils;
+import net.glowstone.net.message.play.entity.UpdateEntityNBTMessage;
+import net.glowstone.util.nbt.CompoundTag;
+
+import java.io.IOException;
+
+public class UpdateEntityNBTCodec implements Codec<UpdateEntityNBTMessage> {
+    @Override
```
Comment (turt2live) [inline src/main/java/net/glowstone/net/codec/play/entity/UpdateEntityNBTCodec.java:14]: Lack of whitespace

File: src/main/java/net/glowstone/net/codec/play/game/ResourcePackSendCodec.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+import com.flowpowered.networking.Codec;
+import com.flowpowered.networking.util.ByteBufUtils;
+import io.netty.buffer.ByteBuf;
+import net.glowstone.net.message.play.game.ResourcePackSendMessage;
+
+import java.io.IOException;
+
+public class ResourcePackSendCodec implements Codec<ResourcePackSendMessage> {
+    @Override
```
Comment (turt2live) [inline src/main/java/net/glowstone/net/codec/play/game/ResourcePackSendCodec.java:12]: lack of whitespace

File: src/main/java/net/glowstone/net/codec/play/game/TitleCodec.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+import com.flowpowered.networking.Codec;
+import com.flowpowered.networking.util.ByteBufUtils;
+import io.netty.buffer.ByteBuf;
+import net.glowstone.net.message.play.game.TitleMessage;
+
+import java.io.IOException;
+
+public class TitleCodec implements Codec<TitleMessage> {
+    @Override
```
Comment (turt2live) [inline src/main/java/net/glowstone/net/codec/play/game/TitleCodec.java:12]: Lack of whitespace

File: src/main/java/net/glowstone/net/codec/play/game/TitleCodec.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    public TitleMessage decode(ByteBuf buffer) throws IOException {
+        TitleMessage titleMessage;
+        TitleMessage.Action action = TitleMessage.Action.getAction(ByteBufUtils.readVarInt(buffer));
+        switch (action) {
+            case TITLE:
+            case SUBTITLE:
+                String text = ByteBufUtils.readUTF8(buffer);
```
Comment (turt2live) [inline src/main/java/net/glowstone/net/codec/play/game/TitleCodec.java:19]: Extra, unrequired, variable

File: src/main/java/net/glowstone/net/codec/play/game/WorldBorderCodec.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+import com.flowpowered.networking.Codec;
+import com.flowpowered.networking.util.ByteBufUtils;
+import io.netty.buffer.ByteBuf;
+import io.netty.handler.codec.DecoderException;
+import net.glowstone.net.message.play.game.WorldBorderMessage;
+
+import java.io.IOException;
+
+public class WorldBorderCodec implements Codec<WorldBorderMessage> {
+    @Override
```
Comment (turt2live) [inline src/main/java/net/glowstone/net/codec/play/game/WorldBorderCodec.java:13]: Lack of whitespace

File: src/main/java/net/glowstone/net/codec/play/player/CameraCodec.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+import com.flowpowered.networking.Codec;
+import com.flowpowered.networking.util.ByteBufUtils;
+import io.netty.buffer.ByteBuf;
+import net.glowstone.net.message.play.player.CameraMessage;
+
+import java.io.IOException;
+
+public class CameraCodec implements Codec<CameraMessage> {
+    @Override
```
Comment (turt2live) [inline src/main/java/net/glowstone/net/codec/play/player/CameraCodec.java:12]: Lack of whitespace

File: src/main/java/net/glowstone/net/codec/play/player/PlayerListHeaderFooterCodec.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+import com.flowpowered.networking.Codec;
+import com.flowpowered.networking.util.ByteBufUtils;
+import io.netty.buffer.ByteBuf;
+import io.netty.handler.codec.DecoderException;
+import net.glowstone.net.message.play.player.PlayerListHeaderFooterMessage;
+
+import java.io.IOException;
+
+public class PlayerListHeaderFooterCodec implements Codec<PlayerListHeaderFooterMessage> {
+    @Override
```
Comment (turt2live) [inline src/main/java/net/glowstone/net/codec/play/player/PlayerListHeaderFooterCodec.java:13]: Lack of whitespace

File: src/main/java/net/glowstone/net/message/play/game/TitleMessage.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        return "TitleMessage{" +
+                "action=" + action +
+                ", text='" + text + '\'' +
+                ", fadeIn=" + fadeIn +
+                ", stay=" + stay +
+                ", fadeOut=" + fadeOut +
+                '}';
+    }
+
+    public enum Action {
```
Comment (turt2live) [inline src/main/java/net/glowstone/net/message/play/game/TitleMessage.java]: Oracle wants enums at the top of classes

File: src/main/java/net/glowstone/net/message/play/game/WorldBorderMessage.java [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+                ", speed=" + speed +
+                ", x=" + x +
+                ", z=" + z +
+                ", portalTeleportBoundary=" + portalTeleportBoundary +
+                ", warningTime=" + warningTime +
+                ", warningBlocks=" + warningBlocks +
+                '}';
+    }
+
+    public enum Action {
```
Comment (turt2live) [inline src/main/java/net/glowstone/net/message/play/game/WorldBorderMessage.java]: Oracle wants enums at the top of files

File: src/main/java/net/glowstone/net/message/play/game/WorldBorderMessage.java [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+        SET_SIZE(0),
+        LERP_SIZE(1),
```
Comment (turt2live) [inline src/main/java/net/glowstone/net/message/play/game/WorldBorderMessage.java]: `LERP` - what the hell is that?

File: src/main/java/net/glowstone/net/message/play/player/CombatEventMessage.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        return "CombatEventMessage{" +
+                "event=" + event +
+                ", duration=" + duration +
+                ", entityID=" + entityID +
+                ", playerID=" + playerID +
+                ", message='" + message + '\'' +
+                '}';
+    }
+
+    public enum Event {
```
Comment (turt2live) [inline src/main/java/net/glowstone/net/message/play/player/CombatEventMessage.java]: Oracle wants enums at the top of the file

File: src/main/java/net/glowstone/net/message/play/game/WorldBorderMessage.java (hunk 10 above)
Comment (greatman) [inline src/main/java/net/glowstone/net/message/play/game/WorldBorderMessage.java]: It's what's written on http://wiki.vg , idk much more

Comment (SpaceManiac) [inline src/main/java/net/glowstone/net/message/play/game/WorldBorderMessage.java]: Linear interpolation, I don't know the details but its job is to specify how fast the world border is moving.

Comment (ZachBora) [inline src/main/java/net/glowstone/net/message/play/game/WorldBorderMessage.java]: Longest Expected Repeated Pattern

**Comment by turt2live** [comment]
Ticket: #131

**Comment by turt2live** [comment]
Sorry - bot fail :(

File: src/main/java/net/glowstone/net/codec/play/game/TitleCodec.java (hunk 4 above)
Comment (daboross) [inline src/main/java/net/glowstone/net/codec/play/game/TitleCodec.java:19]: Looking at this, and just wondering, aren't variables supposed to be like this according to https://github.com/GlowstoneMC/Glowstone/wiki/Code-Style? "In the message codecs, each read from the buffer should be assigned to a variable before the message is constructed."

File: src/main/java/net/glowstone/net/codec/play/game/TitleCodec.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+import net.glowstone.net.message.play.game.TitleMessage;
+
+import java.io.IOException;
+
+public class TitleCodec implements Codec<TitleMessage> {
+
+    @Override
+    public TitleMessage decode(ByteBuf buffer) throws IOException {
+        TitleMessage titleMessage;
+        TitleMessage.Action action = TitleMessage.Action.getAction(ByteBufUtils.readVarInt(buffer));
```
Comment (ZachBora) [inline src/main/java/net/glowstone/net/codec/play/game/TitleCodec.java:15]: missing a separate variable

"In the message codecs, each read from the buffer should be assigned to a variable before the message is constructed."

File: src/main/java/net/glowstone/net/codec/play/game/WorldBorderCodec.java [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+import net.glowstone.net.message.play.game.WorldBorderMessage;
+
+import java.io.IOException;
+
+public class WorldBorderCodec implements Codec<WorldBorderMessage> {
+
+    @Override
+    public WorldBorderMessage decode(ByteBuf buffer) throws IOException {
+        WorldBorderMessage worldBorderMessage;
+        WorldBorderMessage.Action action = WorldBorderMessage.Action.getAction(ByteBufUtils.readVarInt(buffer));
```
Comment (ZachBora) [inline src/main/java/net/glowstone/net/codec/play/game/WorldBorderCodec.java:16]: missing a separate variable

"In the message codecs, each read from the buffer should be assigned to a variable before the message is constructed."

File: src/main/java/net/glowstone/net/codec/play/game/TitleCodec.java (hunk 12 above)
Comment (greatman) [inline src/main/java/net/glowstone/net/codec/play/game/TitleCodec.java:15]: It's exactly doing that. Action is a enum.

File: src/main/java/net/glowstone/net/codec/play/game/WorldBorderCodec.java (hunk 13 above)
Comment (greatman) [inline src/main/java/net/glowstone/net/codec/play/game/WorldBorderCodec.java:16]: It's exactly doing that. Action is a enum.

File: src/main/java/net/glowstone/net/message/play/game/TitleMessage.java [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+            }
+            return null;
+        }
+    }
+
+    private final Action action;
+    private final String text;
+    private final int fadeIn, stay, fadeOut;
+
+
```
Comment (Wild1145) [inline src/main/java/net/glowstone/net/message/play/game/TitleMessage.java]: White space here and line 93

**Comment by turt2live** [comment]
@greatman This does not seem to compile. Please investigate and test your PR. Thanks!

**Comment by dequis** [comment]
Checkstyle failures lol

**Comment by greatman** [comment]
Fixed the checkstyle errors.

**Comment by SpaceManiac** [comment]
Working on getting this merged. I've got the important stuff down but I need confirmation from you on the `CombatEventCodec` - your implementation mismatches the [protocol docs](http://wiki.vg/Protocol#Combat_Event) in several ways. I can take care of fixing this while merging, but I wanted to ask you to figure out whether your current code or the documentation is correct.

**Comment by greatman** [comment]
Hmm I followed the doc. Ill check what's wrong.

**Comment by SpaceManiac** [comment]
Merged in 58e3bf07ad22fd90abf3dd7824c9e34496423df2.

(CombatEventCodec was resolved).

**Pull request closed without merging**